contracts/destination/xreceivers/AuthForwarderXReceiver.sol

```diff
@@ -0,0 +1,229 @@
+   * @param _callData - The data to be prepared and forwarded
+   */
+  function xReceive(
+    bytes32 _transferId,
+    uint256 _amount, // Final Amount receive via Connext(After AMM calculation)
+    address _asset,
+    address _originSender,
+    uint32 _origin,
+    bytes memory _callData
+  ) external onlyOrigin(_originSender, _origin) returns (bytes memory) {
+    // Decode calldata
+    (address _fallbackAddress, bytes memory _data) = abi.decode(_callData, (address, bytes));
+
+    bool successfulForward;
+    try this.prepareAndForward(_transferId, _data, _amount, _asset) returns (bool success) {
+      successfulForward = success;
+      if (!success) {
+        emit ForwardedFunctionCallFailed(_transferId);
+      }
+      // transfer to fallback address if forwardFunctionCall fails
+    } catch Error(string memory _errorMessage) {
+      // This is executed in case
+      // revert was called with a reason string
+      successfulForward = false;
+      emit ForwardedFunctionCallFailed(_transferId, _errorMessage);
+    } catch Panic(uint _errorCode) {
+      // This is executed in case of a panic,
+      // i.e. a serious error like division by zero
+      // or overflow. The error code can be used
+      // to determine the kind of error.
+      successfulForward = false;
+      emit ForwardedFunctionCallFailed(_transferId, _errorCode);
+    } catch (bytes memory _lowLevelData) {
+      // This is executed in case revert() was used.
+      successfulForward = false;
+      emit ForwardedFunctionCallFailed(_transferId, _lowLevelData);
+    }
+    if (!successfulForward) {
+      IERC20(_asset).transfer(_fallbackAddress, _amount);
+    }
+    // Return the success status of the forwardFunctionCall
+    return abi.encode(successfulForward);
+  }
+
+  /// INTERNAL
+  /**
+   * @notice Prepares the data for the function call and forwards it. This can execute
+   * any arbitrary function call in a two step process. For example, _prepare can be used to swap funds
+   * on a DEX, and _forwardFunctionCall can be used to call a contract with the swapped funds.
+   * @dev This function is intended to be called by the xReceive function, and should not be called outside
+   * of that context. The function is `public` so that it can be used with try-catch.
+   *
+   * @param _transferId - The transfer ID of the transfer that triggered this call
+   * @param _data - The data to be prepared
+   * @param _amount - The amount of funds received in this transfer
+   * @param _asset - The asset of the funds received in this transfer
+   */
+  function prepareAndForward(
+    bytes32 _transferId,
+    bytes memory _data,
+    uint256 _amount,
+    address _asset
+  ) public returns (bool) {
+    if (msg.sender != address(this)) {
+      revert ForwarderXReceiver__prepareAndForward_notThis(msg.sender);
+    }
+    // Prepare for forwarding
+    bytes memory _prepared = _prepare(_transferId, _data, _amount, _asset);
```

Comment:
we should likely emit a `Prepared` event here -- users could need this information to debug their transfers since the `_prepare` function is implemented by them